File: src/main/java/hzt/aoc/day15/Part1RambunctiousRecitation.java
package hzt.aoc.day15;

import java.util.List;

public class Part1RambunctiousRecitation extends Day15Challenge {

    public Part1RambunctiousRecitation() {
        super("part 1",
                "Given your starting numbers, what will be the 2020th number spoken?");
    }

    private static final int THRESHOLD = 2020;

    @Override
    protected int getNthNumberSpoken(List<Integer> numbers) {
        int last = 0;
        long start = System.nanoTime();
        while (numbers.size() < THRESHOLD) {
            int prevLast = numbers.remove(numbers.size() - 1);
            int newLast = 0;
            if (numbers.contains(prevLast)) {
                for (int index = numbers.size() - 1; index >= 0; index--) {
                    if (numbers.get(index) == prevLast) {
                        newLast = numbers.size() - index;
                        break;
                    }
                }
            }
            numbers.add(prevLast);
            last = newLast;
            numbers.add(last);
            start = logTime(numbers.size(), 200, 20, last, start);
        }
        return last;
    }

    @Override
    protected String getMessage(String answer) {
        return String.format("The %dth number spoken is: %s", THRESHOLD, answer);
    }

}

File: src/main/java/hzt/aoc/day15/Part1RambunctiousRecitation.kt
package hzt.aoc.day15

import java.lang.System

class Part1RambunctiousRecitation : Day15Challenge(
    "part 1",
    "Given your starting numbers, what will be the 2020th number spoken"
) {
    override fun getNthNumberSpoken(numbers: MutableList<Int>): Int {
        var last = 0
        var start = System.nanoTime()
        while (numbers.size < THRESHOLD) {
            val prevLast = numbers.removeAt(numbers.size - 1)
            var newLast = 0
            if (numbers.contains(prevLast)) {
                for (index in numbers.indices.reversed()) {
                    if (numbers[index] == prevLast) {
                        newLast = numbers.size - index
                        break
                    }
                }
            }
            numbers.add(prevLast)
            last = newLast
            numbers.add(last)
            start = logTime(numbers.size, 200, 20, last, start)
        }
        return last
    }

    override fun getMessage(result: String): String {
        return String.format("The %dth number spoken is: %s", THRESHOLD, result)
    }

    companion object {
        private const val THRESHOLD = 2020
    }
}
